File: digidis-go/day10/day10.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

type point struct {
	x, y, dx, dy int
}

func main() {
	data, _ := ioutil.ReadFile("input.txt")
	rows := strings.Split(string(data), "\n")
	var (
		x, y, dx, dy int
		points       []*point
	)
	for _, r := range rows {
		r = strings.Replace(r, " ", "", -1)
		fmt.Sscanf(r, "position=<%d,%d>velocity=<%d,%d>", &x, &y, &dx, &dy)
		points = append(points, &point{
			x, y, dx, dy,
		})
	}

	second := 1
	centered := false
	for {
		minx, maxx, miny, maxy := 0, 0, 0, 0
		for _, p := range points {
			p.x += p.dx
			p.y += p.dy
			if minx == 0 || p.x < minx {
				minx = p.x
			}
			if p.x > maxx {
				maxx = p.x
			}
			if miny == 0 || p.y < miny {
				miny = p.y
			}
			if p.y > maxy {
				maxy = p.y
			}
		}
		if maxy-miny < 30 {
			centered = true
			fmt.Printf("[Second %d] Min x: %d, Max x: %d, min y: %d, max y: %d => \n", second, minx, maxx, miny, maxy)
			for y := miny; y <= maxy; y++ {
				for x := minx; x <= maxx; x++ {
					found := false
					for _, p := range points {
						if p.x == x && p.y == y {
							found = true
							break
						}
					}
					if found {
						fmt.Printf("#")
					} else {
						fmt.Printf(".")
					}
				}
				fmt.Printf("\n")
			}
		} else {
			if centered {
				return
			}
		}
		second++
	}
}

File: src/adventofcode/2018/day10.clj
(ns adventofcode.2018.day10
  (:require clojure.string))

(def example-input "position=< 9,  1> velocity=< 0,  2>
position=< 7,  0> velocity=<-1,  0>
position=< 3, -2> velocity=<-1,  1>
position=< 6, 10> velocity=<-2, -1>
position=< 2, -4> velocity=< 2,  2>
position=<-6, 10> velocity=< 2, -2>
position=< 1,  8> velocity=< 1, -1>
position=< 1,  7> velocity=< 1,  0>
position=<-3, 11> velocity=< 1, -2>
position=< 7,  6> velocity=<-1, -1>
position=<-2,  3> velocity=< 1,  0>
position=<-4,  3> velocity=< 2,  0>
position=<10, -3> velocity=<-1,  1>
position=< 5, 11> velocity=< 1, -2>
position=< 4,  7> velocity=< 0, -1>
position=< 8, -2> velocity=< 0,  1>
position=<15,  0> velocity=<-2,  0>
position=< 1,  6> velocity=< 1,  0>
position=< 8,  9> velocity=< 0, -1>
position=< 3,  3> velocity=<-1,  1>
position=< 0,  5> velocity=< 0, -1>
position=<-2,  2> velocity=< 2,  0>
position=< 5, -2> velocity=< 1,  2>
position=< 1,  4> velocity=< 2,  1>
position=<-2,  7> velocity=< 2, -2>
position=< 3,  6> velocity=<-1, -1>
position=< 5,  0> velocity=< 1,  0>
position=<-6,  0> velocity=< 2,  0>
position=< 5,  9> velocity=< 1, -2>
position=<14,  7> velocity=<-2,  0>
position=<-3,  6> velocity=< 2, -1>")

(defn parse-input [lines]
  (map (fn [line]
         (->> line
              (re-matches #".*?(-?\d+).*?(-?\d+).*?(-?\d+).*?(-?\d+).*?")
              (rest)
              (map read-string)
              (split-at 2)
              (map #(apply vector %))
              (apply vector)
              ))
       lines
       ))

(defn step [points]
  (map (fn [[[x y] [vx vy]]]
         [[(+ x vx) (+ y vy)] [vx vy]]
         )
       points
       ))

(defn minmax [points]
  (let [pointset (->> points
                      (map first)
                      (set)
                      )]
    [(apply min (map first pointset))
     (apply min (map second pointset))
     (apply max (map first pointset))
     (apply max (map second pointset))
     pointset
     ]))

(defn format-state [points]
  (let [[minx miny maxx maxy pointset] (minmax points)]
    (->> (range miny (inc maxy))
         (map (fn [y]
                (->> (range minx (inc maxx))
                     (map (fn [x]
                            (if (pointset [x y]) \# \.)
                            ))
                     )))
         (map clojure.string/join)
         (clojure.string/join \newline)
         )))

(defn area [points]
  (let [[minx miny maxx maxy] (minmax points)]
    (* (- maxx minx) (- maxy miny))
    ))

(defn solve-a [lines]
  (->> lines
       (parse-input)
       (iterate step)
       (partition 2 1)
       (some (fn [[st1 st2]]
                 (if (< (area st1) (area st2))
                   st1
                   nil
                   )))
       (format-state)
       ))

(defn solve-b [lines]
  ())

(defn run [input-lines & args]
  {:A (solve-a input-lines)
   :B (solve-b input-lines)
   }
  )

(defn run-print []
  (println (solve-a (day-lines)))
  )

(defn day-lines [] (adventofcode.2018.core/day-lines 10))
